text/0000-reactive-assignments.md

```diff
@@ -0,0 +1,946 @@
+  {#if visible}
+    <p>now you see me</p>
+  {/if}
+</div>
+```
+
+```js
+import Peekaboo from './Peekaboo.html';
+
+const peekaboo = new Peekaboo(...);
+peekaboo.show();
+```
+
+An `export const` would be a signal to the compiler that a property is read-only — in other words, attempting to assign to it would cause an error.
+
+
+### `preload` and `setup`
+
+There is a rather awkward mechanism for declaring static properties on a component constructor in Svelte 2 — the `setup` hook:
+
+```html
+<script>
+  export default {
+    setup(Component) {
+      Component.staticMethod = () => {...};
+    }
+  };
+</script>
+```
+
+This is deeply weird, and due to an oversight (that we can't correct without a breaking change) only runs on the client.
+
+Since Sapper requires that components have some way of declaring their data dependencies prior to rendering, and since `setup` is so cumbersome, there is a special case made for `preload`. The `preload` function is attached to components on both client and server, and has no well-defined behaviour; it is purely convention.
+
+We can do better, **but it requires something potentially controversial** — a second `<script>` block that, unlike the one we've been using so far, runs a single time rather than upon every instantiation. Let's call it (🐃) `scope="shared"`:
+
+```html
+<script scope="shared">
+  export function preload({ params }) {
+    return this.fetch(`data/${params.id}.json`).then(r => r.json()).then(things => {
+      return { things };
+    });
+  }
+</script>
+
+<script>
+  export let things;
+</script>
+
+{#each things as thing}
+  <p>{thing}</p>
+{/each}
+```
+
+`things` would be injected into the instance `<script>` block at the same time as props (i.e. between instantiation and first render).
+
+
+### Server-side rendering
+
+The compiler running with the `generate: 'ssr'` option produces completely different code from `generate: 'dom'`. This is because server-rendered components do not have a lifespan — they are rendered, then immediately discarded, so rather than having a concept of 'instances' an SSR component exports a function that given some data creates HTML and CSS.
+
+This does create some subtle behaviour differences however: there is no real place to do any kind of setup work, and `oncreate` (and the other lifecycle hooks) never run. Component bindings are also brittle.
+
+Under this proposal, the code inside the `<script>` block *would* run for server-rendered components, which also means that `ondestroy` would need to run. `onstate` and `onupdate` would be no-ops.
+
+I don't think the API needs to change.
+
+
+### Standalone components
+
+It's not clear how this proposal affects standalone components (as opposed to those that use shared helpers). Perhaps it doesn't?
+
+One consequence of the new way of doing things: Svelte no longer lives in `pkg.devDependencies`, but would instead need to become a `dependency`. Not very On Brand but probably a small price to pay.
```

Comment:
As I noted in chat, I don't see why this is an issue. It seems like we could just inline these `import { whatever } from 'svelte'` statements, and avoid the dependence on `svelte` at runtime

Reply:
yeah fair enough. removed. side-note — I was thinking that it probably makes sense to just use Rollup to create the standalone things (with everything except `svelte` considered external), rather than the current Heath Robinson nonsense

Reply:
That sounds way saner than the current mechanisms for this, but I am a bit worried about the compiler's size. Am I going to be forced to download Svelte with a bundled-in version of Rollup? Am I going to be forced to install it as a `dependency` even if (for some reason) I don't want it? (Or if I already have a different/newer version of Rollup as a dependency of my app?)

Do we want to make Rollup an optionalDependency, and throw if we try to compile in standalone mode with it unavailable?

Reply:
definitely a fair concern. also, it would force `svelte.compile` to become async. maybe we should cross that bridge later

Reply:
Do we want to _drop_ standalone component compilation as part of `svelte.compile` and introduce a separate function that does require Rollup, or maybe some sort of drop-in Rollup plugin or config?

One possibility would be to make this not `svelte.compile`'s concern at all, and to take care of it in the `component-template` `rollup.config.js`.

I do see the value of no longer having to worry about this weirdness, but I am concerned that we will be stuck with it until Svelte v4 unless we do _something_ as part of these changes.

Reply:
I think we should strongly consider it. I don't think it needs to make it into *this* RFC (it doesn't need to cover every aspect of v3) but yeah, we should try not to lose track of it. Standalone components are a useful selling point of this framework but they are also a total bloody nuisance

Reply:
Wait I just realized, this would also probably let us leave behind all of the non-ESM output modes too. Dang this is sounding real tasty now.

Reply:
Ugh I just remembered the Svelte CLI which people are still using for some reason. I don't know how that fits into this anymore. Do we tell the people who were using it to create standalone IIFEs to get lost?